### PR TITLE
Fix #1003 - permit input utf8

### LINKED ISSUE #1003: Internal IDE utf-8 support
So, the game already translated to many languages, but IDE supports only `latin`.
Any chance for `utf-8`?

### PR DESCRIPTION
**Pull request by B-CE**
fix #1003

Warning  (see #1003 comments) :
it will put in evidence some issues, because lots of checks use the "ANSI char encoding" into CEdit class : 
word detection, missing chars in lines, carret moves, 2 or 3 backspace needed to delete one char, ...

(#1114 embedded to permit compilation)

### CODE REVIEW

**Comment by B-CE** [comment]
Even if the characters are now internally managed, the display isn't still completely correct :
(also maybe linked to "dvu_sans.ttf" choice )

The application seams well setted (see app/app.cpp )

 
samples:
OK :
 éçàηџŎЄ ħëłlö    œ ž Ÿ ų  ě γλώσα   Ɣ Ƨ  ʨ ʬ ͡ Җ Ӌ ն ٢  
Bulgarian: Мога да ям стъкло и не ме боли.
Russian: 	Я могу есть стекло, это мне не вредит.
Serbian: 	Могу јести стакло а да ми не шкоди.

partial Greek & Russian

KO : chinese, japanese, arabic ...
€ Ѥ⤕ԋセ⸂̸⸒你好国家标准     ě出力 𐎮 𐍉 Ѡ ҉ ؄  ٯ  ټ  ₡ ₢ ₣ ₤ ₥ ₦ ₧ ₨ ₩ ₪ ₫ ₭ ₮ ₯ ₹


(btw, even some of my tested characters are not displayed there ^^ :
to sum up: most of 1 & 2 bytes works, quasi all 3 & 4 bytes aren't correctly displayed ... )

Greek Monotonic OK (Some lines of Odysseus Elytis )
Τη γλώσσα μου έδωσαν ελληνική
το σπίτι φτωχικό στις αμμουδιές του Ομήρου.
Μονάχη έγνοια η γλώσσα μου στις αμμουδιές του Ομήρου


Greek Polytonic not fully OK (Same lines of Odysseus Elytis )
Τὴ γλῶσσα μοῦ ἔδωσαν ἑλληνικὴ
τὸ σπίτι φτωχικὸ στὶς ἀμμουδιὲς τοῦ Ὁμήρου.
Μονάχη ἔγνοια ἡ γλῶσσα μου στὶς ἀμμουδιὲς τοῦ Ὁμήρου


Russian seams OK
На берегу пустынных волн
Стоял он, дум великих полн,
И вдаль глядел. Пред ним широко
..

 Anglo-Saxon Rune Poem (Rune version):  are KO ... but honnestly, who use it into code ^^ ?... :
ᚠᛇᚻ᛫ᛒᛦᚦ᛫ᚠᚱᚩᚠᚢᚱ᛫ᚠᛁᚱᚪ᛫ᚷᛖᚻᚹᛦᛚᚳᚢᛗ
ᛋᚳᛖᚪᛚ᛫ᚦᛖᚪᚻ᛫ᛗᚪᚾᚾᚪ᛫ᚷᛖᚻᚹᛦᛚᚳ᛫ᛗᛁᚳᛚᚢᚾ᛫ᚻᛦᛏ᛫ᛞᚫᛚᚪᚾ
ᚷᛁᚠ᛫ᚻᛖ᛫ᚹᛁᛚᛖ᛫ᚠᚩᚱ᛫ᛞᚱᛁᚻᛏᚾᛖ᛫ᛞᚩᛗᛖᛋ᛫ᚻᛚᛇᛏᚪᚾ᛬

**Comment by B-CE** [comment]
now most seam ok.
Europeans, Russian, Greek even polytonic, Arabic,  Armenian...

and some are still KO:
 Chinese, Japanese, Hebrew ... & runes ^^

http://generator.lorem-ipsum.info/_russian
http://www.utf8-chartable.de/unicode-utf8-table.pl?start=1536&number=1024

**Comment by B-CE** [comment]
may also fix #477

**Comment by B-CE** [comment]
( improvement inwork, please don't waste time checking it yet )

File: lib/gtest/include/gtest/gtest-param-test.h [hunk 1]
```diff
@@ -314,7 +314,7 @@ ValuesIn(ForwardIterator begin, ForwardIterator end) {
       new internal::ValuesInIteratorRangeGenerator<ParamType>(begin, end));
 }
 
-template <typename T, size_t N>
+template <typename T, std::size_t N>
```
Comment (krzys-h) [inline lib/gtest/include/gtest/gtest-param-test.h]: It looks like you changed library files by mistake, please revert that

File: src/graphics/engine/text.cpp [hunk 2]
```diff
@@ -780,32 +874,46 @@ void CText::DrawString(const std::string &text, std::vector<FontMetaChar>::itera
...
+    std::size_t totalLength = text.length();
     while (index < totalLength)
     {
         UTF8Char ch;
 
-        int len = StrUtils::Utf8CharSizeAt(text, index);
+        short len = StrUtils::Utf8CharSizeAt(text, index);
+        if (len == 0)   //warning & skip all : Risk of infinite loop !!
+        {
+            GetLogger()->Warn("Risk of infinite loop 6 -CText::StringToUTFCharList !!\n");
```
Comment (krzys-h) [inline src/graphics/engine/text.cpp]: This code with warnings is repeating all the time, how about extracting it to a function?

File: src/common/stringutils.cpp [hunk 3]
```diff
@@ -24,6 +24,8 @@
 #include <cstdio>
 #include <vector>
 
+#include <cstring>          // memset used for trace
+#include "common/logger.h"
```
Comment (krzys-h) [inline src/common/stringutils.cpp]: I reenabled our old linter and it complains about include order here, local includes should go before global ones

File: lib/gtest/include/gtest/gtest-param-test.h (hunk 1 above)
Comment (B-CE) [reply lib/gtest/include/gtest/gtest-param-test.h]: Sure. I'd applied a regexp to respect the codings rules...

**Comment by B-CE** [comment]
Asked changes had been impl.

Those Jenkins' merge are strange, & I'm preparing an other big style fix 
 
Please just check following fix during conflict resolution: (   in doubt, both parts  concat )

```
void CEngine::AddBaseObjQuick

<<<<<<< dev-inputUTF8
        p1.radius = Math::Max(p1.bboxMin.Length(), p1.bboxMax.Length());
=======
        p1.boundingSphere = Math::BoundingSphereForBox(p1.bboxMin, p1.bboxMax);
>>>>>>> dev

```

**Comment by krzys-h** [comment]
Honestly, there is no way 8000 lines of code changes are ever going to be merged, nobody is going to be able to review that. This needs to be redone without thousands of unrelated and formatting changes and littering the whole code with Unicode checks, instead abstracting them away to one place. The amount of random formatting changes in this PR makes it impossible to even merge and test now because the changes conflict with literally everything.

**Comment by B-CE** [comment]
Well, ok, understood.
That plus the merge was a mistake. It'll be redone properly.

**Pull request closed without merging**